Draft: Add RedirectToActionOther method.

minecraftchest1 opened the pull request:
# Add RedirectToActionOther Method.

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Summary of the changes (Less than 80 chars)

## Description

Adds the method RedirectToActionOther to the class Microsoft.AspNetCore.Mvc that ahould allow redirecting to an action without preserving the HTTP Request Method.

Fixes #36172

minecraftchest1 committed:
Add RedirectToControllerOther method.

mitchdenny commented:
We probably want to take this through API review. Can you update the original issue using this template:

https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=

This helps ensure the right people get eyes on it can can tweak things. Your PR build failed due to some code formatting issues and the fact that the public API text file needs to be updated. If you hover over the new method there should be a Fix in VS that updates this file correctly for you.

minecraftchest1 committed:
Remove extra newlines.

minecraftchest1 commented:
> We probably want to take this through API review. Can you update the original issue using this template:
> 
> https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=
> 
> This helps ensure the right people get eyes on it can can tweak things. Your PR build failed due to some code formatting issues and the fact that the public API text file needs to be updated. If you hover over the new method there should be a Fix in VS that updates this file correctly for you.

I removed the extra newlines from my code. I don't have accessmto Visual Studio (I am runing linux), and VSCode makes my laptop want to fall over dead. Can you point me to file I should update please?

minecraftchest1 committed:
Add RedirectToActionOther to PublicApi.

minecraftchest1 commented:
I believe I made the correct changes to the PublicApi file.

Eilon commented on src/Mvc/Mvc.Core/src/ControllerBase.cs:
```diff
@@ -564,6 +564,123 @@ public virtual RedirectToActionResult RedirectToAction(string? actionName, strin
+    /// public IActionResult Product(int id, Product product)
+    /// {
+    ///     UpdateProduct(product);
+    ///     return RedirectToActionOthet();
```
Typo in the sample code: `othet`

minecraftchest1 replied:
Fixed. Pushing change now.

minecraftchest1 commented:
Thanks. I will fix it when I get off work. I typed it up late last night, so typo's don't supprise me.



Sent from my T-Mobile 4G LTE Device

minecraftchest1 committed:
Fix spelling mistake

Change "Othet" to "Other" in line 584 of src/Mvc/Mvc.Core/src/ControllerBase.cs

minecraftchest1 committed:
Move changes from Shipped to Unshipped.

minecraftchest1 commented:
All requested changes made. What else do I need to do?

mitchdenny commented:
Because this is a change/addition to the public API we need to actually do the API review. When you create a new issue you can select the API proposal template which has all the details that need to be covered. You can either create a new issue or update the original link issue with the contents from the template. Link it to this PR so we can track everything. Then we can look at getting your proposal put through API review.

minecraftchest1 commented:
/azp run



Sent from my T-Mobile 4G LTE Device

minecraftchest1 commented:
/azp run

captainsafia commented:
@minecraftchest1 The `/azp run` command is limited to maintainers at the moment. As mentioned in [this comment](https://github.com/dotnet/aspnetcore/pull/48823#issuecomment-1596438528) you'll want to open an API review issue for this PR first to move to the next step. Do you need help with that?

minecraftchest1 commented:
Yes.



Sent from my T-Mobile 4G LTE Device

captainsafia commented:
You can find the API review template [here](https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=). There are descriptions under each heading of the template of the content that's needed.

aetos382 commented:
This implementation does not resolve the request in #36172.
The result of `RedirectToActionOther` is identical to `RedirectToAction`.
The [RedirectToActionResultExecutor](https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/Infrastructure/RedirectToActionResultExecutor.cs) must be changed to generate a 303 response.

aetos382 commented:
This implementation does not resolve the request in #36172.
The result of RedirectToActionOther is identical to RedirectToAction.

The response generated by `RedirectToActionResultExecutor` is determined by the parameters of `RedirectToActionResult` as follows.

|Method of Controller|preserveMethod|permanent|response|
|-|-|-|-|
|RedirectToAction|false|false|302|
|RedirectToActionPermanent|false|true|301|
|RedirectToActionPreserveMethod|true|false|307|
|RedirectToActionPermanentPreserveMethod|true|true|308|

The problem is that the `preserveMethod` parameter of type `bool` cannot express the semantics that "the request must be made with the GET method".
However, I disagree with the idea of making it an enumerated type representing the three options.
Suppose we create a new type like the following.

```cs
enum RedirectionMethod
{
    AllowChange,
    Preserve,
    MustGet
}
```

In this case, the `permanent` parameter is meaningless for `MustGet`.
For this reason, I think a new Result type and Executor for it should be added to be able to generate a 303 response.

The response codes 301, 302, 307, and 308 all (essentially) mean "The user's request has not been completed and the user agent must retry the request for the given URL."
The 303 response differs from them in that it indicates that "The user's request has been completed and the server is presenting next instruction."
Therefore, I do not believe that 303 should be treated as a variation of other redirects.

codlin commented:
Is the 303 issue still unresolved?
